### PR TITLE
when making source account, randomly sleep between tries 

### PR DESCRIPTION
**Pull request by msmouse**
### Description
1. sleep between tries to make source account, because the failure is possible other instances of the emitter trying to                                                                                                                   use the same sequence number   
2. similarly, for mint_to_root
3. do not retry on SEQUENCE_NUMBER_TOO_OLD -- won't succeed
4. configurable latency-poll-interval (stolen from @igor-aptos )

### CODE REVIEW

**Review by areshand — APPROVED** [review]

File: crates/transaction-emitter-lib/src/emitter/account_minter.rs [hunk 1]
```diff
@@ -380,8 +383,15 @@ impl<'t> AccountMinter<'t> {
                 error!(
                     "Couldn't create new source account, {:?}, try {}, retrying",
                     e, i
-                )
+                );
+                // random sleep to coordinate with other instances
+                std::thread::sleep(Duration::from_secs(rand::thread_rng().gen_range(0, 10)));
             } else {
+                new_source_account.set_sequence_number(
```
Comment (sitalkedia) [inline crates/transaction-emitter-lib/src/emitter/account_minter.rs:390]: This is unrelated?

Comment (msmouse) [reply crates/transaction-emitter-lib/src/emitter/account_minter.rs:390]: that's the base pr https://github.com/aptos-labs/aptos-core/pull/10837

**Review by sitalkedia — APPROVED** [review]

File: config/src/config/api_config.rs [hunk 2]
```diff
@@ -77,7 +77,7 @@ pub struct ApiConfig {
 const DEFAULT_ADDRESS: &str = "127.0.0.1";
 const DEFAULT_PORT: u16 = 8080;
 const DEFAULT_REQUEST_CONTENT_LENGTH_LIMIT: u64 = 8 * 1024 * 1024; // 8 MB
-pub const DEFAULT_MAX_SUBMIT_TRANSACTION_BATCH_SIZE: usize = 10;
+pub const DEFAULT_MAX_SUBMIT_TRANSACTION_BATCH_SIZE: usize = 100;
```
Comment (gregnazario) [inline config/src/config/api_config.rs:80]: This was purposely set to 10, if you want to increase it to 100 for testing please update it in the node configs

File: crates/transaction-emitter-lib/src/emitter/account_minter.rs [hunk 3]
```diff
@@ -277,14 +280,28 @@ impl<'t> AccountMinter<'t> {
...
+                .source_account
+                .sign_with_transaction_builder(self.txn_factory.payload(
+                    aptos_stdlib::aptos_coin_mint(self.source_account.address(), amount),
+                ));
+            match txn_executor.execute_transactions(&[txn]).await {
+                Ok(_) => return Ok(()),
+                Err(e) => {
+                    error!("Couldn't mint to root, {:?}, try {}, retrying", e, i);
+                    // random sleep to coordinate with other instances
+                    std::thread::sleep(Duration::from_secs(rand::thread_rng().gen_range(0, 10)));
```
Comment (gregnazario) [inline crates/transaction-emitter-lib/src/emitter/account_minter.rs:300]: If it's async, you should use Tokio sleep vs thread sleep.  Otherwise, the whole thread's tasks will sleep for that long.  10 seconds is a long time.  If you're creating 10k addresses and this happens every time, you could wait 100k seconds, which is greater than 3 hours

File: crates/transaction-emitter-lib/src/emitter/account_minter.rs [hunk 4]
```diff
@@ -380,7 +397,9 @@ impl<'t> AccountMinter<'t> {
                 error!(
                     "Couldn't create new source account, {:?}, try {}, retrying",
                     e, i
-                )
+                );
+                // random sleep to coordinate with other instances
+                std::thread::sleep(Duration::from_secs(rand::thread_rng().gen_range(0, 10)));
```
Comment (gregnazario) [inline crates/transaction-emitter-lib/src/emitter/account_minter.rs:402]: Similarly

**Review by gregnazario — DISMISSED** [review]
Have we tested this?  I'm a little concerned this could make forge take forever, when this should really be handled by sequence number coordination

File: crates/transaction-emitter-lib/src/emitter/account_minter.rs [hunk 5]
```diff
@@ -277,14 +280,28 @@ impl<'t> AccountMinter<'t> {
         amount: u64,
     ) -> Result<()> {
         info!("Minting new coins to root");
+        for i in 0..3 {
+            self.source_account.set_sequence_number(
+                txn_executor
+                    .query_sequence_number(self.source_account.address())
+                    .await?,
+            );
```
Comment (gregnazario) [inline crates/transaction-emitter-lib/src/emitter/account_minter.rs:289]: A shared DB among instances or some shared state between threads and emitters should keep track of the sequence number.  Right now you must have multiple items fighting over the same transactions.

The transactions should probably also be queued

File: crates/transaction-emitter-lib/src/emitter/account_minter.rs (hunk 3 above)
Comment (msmouse) [reply crates/transaction-emitter-lib/src/emitter/account_minter.rs]: these two positions we do only once per emitter run
good point on async sleep 👍

File: crates/transaction-emitter-lib/src/emitter/account_minter.rs (hunk 5 above)
Comment (msmouse) [reply crates/transaction-emitter-lib/src/emitter/account_minter.rs:290]: these runs from different machines (and only once per run)

**Review by igor-aptos — CHANGES_REQUESTED** [review]
I don't think you need any of this after my PR, right?

**Comment by msmouse** [comment]
@igor-aptos  well, making the source account can still conflict, and waiting and retrying on SEQUENCE_NUMBER_TOO_OLD is always an issue? (And we wait for too long for too many times)

File: crates/transaction-emitter-lib/src/emitter/transaction_executor.rs [hunk 6]
```diff
@@ -99,20 +99,28 @@ impl RestApiReliableTransactionSubmitter {
...
+                            .get(&rest_client.path_prefix_string())
+                            .map(|(successes, _, _)| {
+                                successes.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+                            });
+                    }
+                    return Ok(());
+                },
+                Err(err) => {
+                    // TODO: we should have a better way to decide if a failure is retryable
+                    if format!("{}", err).contains("SEQUENCE_NUMBER_TOO_OLD") {
```
Comment (igor-aptos) [inline crates/transaction-emitter-lib/src/emitter/transaction_executor.rs:119]: I see, so this and the change below. makes sense.

do you want to revert the rest of the PR, so we can land it?

Comment (msmouse) [reply crates/transaction-emitter-lib/src/emitter/transaction_executor.rs:119]: k on it

**Review by igor-aptos — APPROVED** [review]
